### DISCUSSION TITLE
Spinel RCP messages max size and communication mode

### DISCUSSION THREAD

**Discussion by rretanubun** [openthread/openthread#9330] [Q&A]
Hello Everyone,

## Context
Our platform is ESP32 Host Processor + EFR32 RCP.
We are seeing an issue where when there is simultaneous OpenThread access and Flash Writes
we can see a crash.

One of the hypothesis we are pursuing is the use of
CONFIG_UART_ISR_IN_IRAM in ESP-IDF (originally `true`, we are disabling this).

We are aware of this warning from ESP-IDF:
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/kconfig.html#config-uart-isr-in-iram

Essentially, with this disabled, on heavy flash writes activity (e.g. processing an OTA Firmware update over LAN)
there is a chance that some communication between the OT host processor and its RCP could be missed.

## What we need help with
We have a 128 byte FIFO buffer on the RX so we think we can sustain the interruption and service it later, 
provided (`and this is what we need help confirming`):
1. RCP Spinel operates using Request/Response model
2. No RCP message is longer than 128 bytes (because it will overflow the FIFO)

Any hints are much appreciated,

Thanks for everyone's time!

**Comment by zhanglongxia** [comment]
Hope following answers could help you:
1. From the host to the RCP, most of the RCP Spinel commands use the command and response model. But there is an exception, the command `SPINEL_PROP_STREAM_RAW` [won't wait for the response](https://github.com/openthread/openthread/blob/main/src/lib/spinel/radio_spinel_impl.hpp#L1608). 
2. Some Spinel messages may be longer than 128 bytes, such as `SPINEL_PROP_STREAM_RAW`.

**Comment by chshu** [comment]
@rretanubun Do you have any specific reason for disabling `CONFIG_UART_ISR_IN_IRAM`? By placing the UART ISR function into IRAM, the UART receiving ISR can continue to work during flash operations.

Even if some UART data is lost, the host should not crash, as there are retransmission and recovery mechanisms in place. Could you share us the IDF commit and the crash log?

We have tested this scenario on ESP Thread BR devkit: https://docs.espressif.com/projects/esp-thread-br, but didn't see any issue during host OTA.

**Reply by Mikeadaneg** [reply]
@chshu The reason we disabled the CONFIG_UART_ISR_IN_IRAM setting is because we were encountering errors when the UART ISR callbacks were triggered while the PSRAM / flash cache was disabled. By turning off CONFIG_UART_ISR_IN_IRAM, we've seen improvements.
The IDF commit relevant to our discussion can be found here: https://github.com/espressif/esp-idf/tree/6a7d83af1984b93ebaefa7ca9be36304806c3dc8/. Additionally, I have also attached crash log.
[crash log.txt](https://github.com/openthread/openthread/files/12233247/crash.log.txt).
For clarity, we're not using an Espressif ESP-H2 as our RCP.

**Reply by chshu** [reply]
@Mikeadaneg The crash appears to be unrelated to UART data loss, did the crash log happen with `CONFIG_UART_ISR_IN_IRAM` disabled or enabled? Could you provide us the `sdkconfig.defaults` file from your project, we will try to reproduce.

**Reply by Mikeadaneg** [reply]
@chshu, the crash log I shared has the CONFIG_UART_ISR_IN_IRAM setting enabled. Due to the sensitive nature of the sdkconfig which contains customer-specific details, I'd prefer to share it directly through email. Could you please suggest an alternative secure method for me to share the sdkconfig with you.
Thank you for your understanding and cooperation.

**Reply by chshu** [reply]
please send it to chenshu@espressif.com, we will get back to you soon.
